livedata/src/main/java/io/getstream/chat/android/livedata/EventBatchUpdate.kt

```diff
@@ -0,0 +1,109 @@
+package io.getstream.chat.android.livedata
+
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.livedata.extensions.updateLastMessage
+import io.getstream.chat.android.livedata.extensions.users
+
+/**
+ * EventBatchUpdate helps you efficiently implement a 4 step batch update process
+ * It updates multiple messages, users and channels at once.
+ *
+ * val batchBuilder = EventBatchUpdate.Builder()
+ *
+ * as a first step specify which channels and messages to fetch
+ * batchBuilder.addToFetchChannels()
+ * batchBuilder.addToFetchMessages()
+ *
+ * as a second step, load the required data for batch updating using
+ * val batch = batchBuilder.build(domainImpl)
+ *
+ * third, add the required updates via
+ * batch.addUser, addChannel and addMessage methods
+ *
+ * fourth, execute the batch using
+ * batch.execute()
+ */
+internal class EventBatchUpdate private constructor(
+    private val domainImpl: ChatDomainImpl,
+    channelMap: Map<String, Channel>,
+    messageMap: Map<String, Message>
+) {
+    private var users: Map<String, User> = emptyMap()
+    private var channels: Map<String, Channel> = channelMap
+    private var messages: Map<String, Message> = messageMap
+
+    internal class Builder {
+        private var channelsToFetch = setOf<String>()
+        private var messagesToFetch = setOf<String>()
+        private var users = setOf<User>()
+
+        fun addToFetchChannels(cIds: List<String>) {
+            channelsToFetch = channelsToFetch + cIds
```

Comment:
In this case, we're adding _a lot_ of elements to these lists, and we're definitely only doing it on one thread at a time. Keeping these collections mutable and adding to them instead of allocating new ones for each function call would be a good idea.